MAVProxy/modules/mavproxy_chat/__init__.py

```diff
@@ -55,6 +62,53 @@ def cmd_chat(self, args):
     def show(self):
         self.chat_window.show()
 
+    # handle mavlink packet
+    def mavlink_packet(self, m):
+        if m.get_type() == 'COMMAND_ACK':
+            self.handle_command_ack(m)
+
+    # handle_command_ack.  should be called if module receives a COMMAND_ACK command
+    def handle_command_ack(self, m):
+        # return immediately if we are not waiting for this command ack
+        if m.command not in self.command_ack_waiting:
+            return
+
+        # throw away value if result in progress
+        if m.result == mavutil.mavlink.MAV_RESULT_IN_PROGRESS:
+            return
+
+        #  set the mav result for this command
+        self.command_ack_waiting[m.command] = m.result
+
+    # wait for COMMAND_ACK with the specified mav_cmd
+    # this should be called immediately after sending a command_long or command_int
+    # mav_cmd should be set to the command id that was sent
+    # returns MAV_RESULT if command ack received, False if timed out
+    def wait_for_command_ack(self, mav_cmd, timeout=1):
+        # error if we are already waiting for this command ack
+        if mav_cmd in self.command_ack_waiting:
+            print("chat: already waiting for command ack for mavcmd:" + str(mav_cmd))
+            return False
+
+        # add to list of commands we are waiting for (-1 indicates we don't know result yet)
+        self.command_ack_waiting[mav_cmd] = -1
+
+        # wait for ack, checking for a response every 0.1 seconds
+        start_time = time.time()
+        while time.time() - start_time < timeout:
```

Comment:
You can't do this in a lot of contexts as you will kill MAVProxy's main loop.

You should (a) test what happens if you sleep for 10 minutes in here and (b) add to the comment at the top of this method indicating that it's only OK to call this from the chat module's thread or whatever.

Reply:
Hi, in this case the chat window runs in a separate thread so MAVProxy's main loop is not affected.  I've added a comment to the top of the function to specify that it blocks and shouldn't be called from the main thread.

`# Note: this should not be called from the main thread because it blocks for up to "timeout" seconds`

I've also tested the long sleep in this function and it doesn't freeze up MAVProxy.  The chat module window can even still be resized and moved around but it gets stuck for a long time before it provides an update to the user.